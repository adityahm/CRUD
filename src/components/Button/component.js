import React from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

const Button = ({
  type,
  action,
  text,
  icon,
  iconSize,
  className,
  iconButton,
  children
}) => (
  <button
    data-test="button-component"
    type={type}
    onClick={action}
    className={`${iconButton ? 'icon-button' : 'button'} ${className}`}
  >
    {icon ? <FontAwesomeIcon icon={icon} size={iconSize} /> : null}
    {text}
    {children}
  </button>
);

Button.propTypes = {
  type: PropTypes.string,
  action: PropTypes.func.isRequired,
  text: PropTypes.string,
  icon: PropTypes.string,
  iconSize: PropTypes.string,
  className: PropTypes.string,
  iconButton: PropTypes.bool,
  children: PropTypes.node
};

Button.defaultProps = {
  type: 'button',
  className: '',
  text: '',
  iconSize: 'lg',
  iconButton: false,
  action: () => null
};

export default Button;
