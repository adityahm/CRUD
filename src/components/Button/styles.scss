.icon-button {
    border: none;
    outline: none;
    background: none;
    color: inherit;
}