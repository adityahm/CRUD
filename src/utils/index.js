export const findTestWrapper = (component, attr) => {
  const wrapper = component.find(`[data-test='${attr}']`);
  return wrapper;
};
