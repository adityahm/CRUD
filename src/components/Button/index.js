import Button from './component'
import './styles.scss'

export default Button
