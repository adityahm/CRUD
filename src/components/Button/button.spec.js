import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faUserCircle } from '@fortawesome/free-solid-svg-icons';
import Button from './component';
import { findTestWrapper } from '../../utils';

library.add(faUserCircle);

describe('Button Component', () => {
  it('matches the snapshot', () => {
    const tree = renderer
      .create(
        <Button
          action={() => null}
          type="submit"
          text="Submit"
          icon="user-circle"
          iconSize="lg"
          iconButton
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('button renders with text', () => {
    const wrapper = shallow(<Button text="Submit" action={() => null} />);

    expect(findTestWrapper(wrapper, 'button-component').text()).toBe('Submit');
  });

  it('button action is called on click', () => {
    const mockFn = jest.fn();
    const wrapper = shallow(<Button text="Click" action={mockFn} />);

    wrapper.simulate('click');

    expect(mockFn.mock.calls.length).toBe(1);
  });

  it('button renders with proper type', () => {
    const wrapper = shallow(
      <Button type="reset" text="Click" action={() => null} />
    );

    expect(findTestWrapper(wrapper, 'button-component').prop('type')).toBe(
      'reset'
    );
  });

  it('button renders children', () => {
    const wrapper = shallow(
      <Button type="reset" text="Click" action={() => null}>
        <span>#</span>
      </Button>
    );

    expect(
      findTestWrapper(wrapper, 'button-component').contains(<span>#</span>)
    ).toBeTruthy();
  });
});
