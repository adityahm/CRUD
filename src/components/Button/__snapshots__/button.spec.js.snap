// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches the snapshot 1`] = `
<button
  className="icon-button "
  data-test="button-component"
  onClick={[Function]}
  type="submit"
>
  Submit
</button>
`;
